### PR TITLE
feat(cli): implementation of the new bootstrap logic

### PR DESCRIPTION
**Pull request by skinny85**
This is the logic of the new bootstrap command, deploying the template discussed in #4461 and then further in #5757 . You activate the new bootstrap behavior by passing the `--new-bootstrap=true` option to the CLI.

I've manually verified that bootstrapping succeeds, and that a CDK app using assets can successfully be deployed using this new bootstrap stack.

I've posting this as a draft, to get early feedback, as there are a few ways we can go about incorporating this new functionality, so I wanted to get the team's opinion on the direction taken.

I'm in the process of writing an automated integ test that checks whether this works without the need for manual testing, but it's proving to be a little tricky, so I figured I won't hold up the main logic PR on it.

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/aws-cdk/bin/cdk.ts [hunk 1]
```diff
@@ -56,6 +57,7 @@ async function parseCommandLineArguments() {
       .option('bootstrap-kms-key-id', { type: 'string', desc: 'AWS KMS master key ID used for the SSE-KMS encryption', default: undefined })
       .option('tags', { type: 'array', alias: 't', desc: 'Tags to add for the stack (KEY=VALUE)', nargs: 1, requiresArg: true, default: [] })
       .option('execute', {type: 'boolean', desc: 'Whether to execute ChangeSet (--no-execute will NOT execute the ChangeSet)', default: true})
+      .option('new-bootstrap', { type: 'boolean', desc: 'Whether to use the experimental new enhanced bootstrapping needed for the "CI/CD for CDK apps" story. Default is the classic bootstrapping', default: false })
```
Comment (rix0rrr) [inline packages/aws-cdk/bin/cdk.ts]: It needs to be an option, but I wonder if we should hide it more. Either disclaim it more here ("will change in the future, experimental, don't use etc tec") or maybe make it an environment variable that toggles this?

File: packages/aws-cdk/lib/api/bootstrap/bootstrap-template.json [hunk 2]
```diff
@@ -0,0 +1,394 @@
...
+          "Fn::Equals": [
+            "",
+            {
+              "Ref": "FileAssetsBucketName"
+            }
+          ]
+        }
+      ]
+    },
+    "CustomKmsKeyIdNotPassed": {
```
Comment (rix0rrr) [inline packages/aws-cdk/lib/api/bootstrap/bootstrap-template.json]: `"CreateNewKey"` ?

File: packages/aws-cdk/lib/api/bootstrap/bootstrap-template.json [hunk 3]
```diff
@@ -0,0 +1,394 @@
...
+      "Condition": "CustomKmsKeyIdNotPassed"
+    },
+    "StagingBucket": {
+      "Type": "AWS::S3::Bucket",
+      "Properties": {
+        "BucketName": {
+          "Fn::If": [
+            "HasCustomFileAssetsBucketName",
+            { "Fn::Sub": "${FileAssetsBucketName}" },
+            { "Fn::Sub": "cdk-bootstrap-hnb659fds-assets-${AWS::AccountId}-${AWS::Region}" }
```
Comment (rix0rrr) [inline packages/aws-cdk/lib/api/bootstrap/bootstrap-template.json]: hnb659fds?

File: packages/aws-cdk/lib/api/bootstrap/bootstrap-template.json [hunk 4]
```diff
@@ -0,0 +1,394 @@
...
+      },
+      "Condition": "HasTrustedPrincipals"
+    }
+  },
+  "Outputs": {
+    "BucketName": {
+      "Description": "The name of the S3 bucket owned by the CDK toolkit stack",
+      "Value": { "Fn::Sub":  "${StagingBucket}" },
+      "Export": {
+        "Name": { "Fn::Sub": "${AWS::StackName}:BucketName" }
```
Comment (rix0rrr) [inline packages/aws-cdk/lib/api/bootstrap/bootstrap-template.json]: Do we strictly need these exports?

File: packages/aws-cdk/bin/cdk.ts (hunk 1 above)
Comment (eladb) [reply packages/aws-cdk/bin/cdk.ts]: +1 for environment variable

File: packages/aws-cdk/lib/api/bootstrap/bootstrap-environment2.ts [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+  });
+
+  const assembly = builder.buildAssembly();
+  return await deployStack({
+    stack: assembly.getStackByName(toolkitStackName),
+    sdk,
+    roleArn,
+    tags: props.tags,
+    execute: props.execute
+  });
```
Comment (eladb) [inline packages/aws-cdk/lib/api/bootstrap/bootstrap-environment2.ts]: Perhaps extract this to `deployStackFromTemplate` as a more low-level building block and then just call it from here with the bootstrap template.

Also, I expected to see the bootstrap options passed in to CFN parameters...

File: packages/aws-cdk/lib/api/bootstrap/bootstrap-template.json [hunk 6]
```diff
@@ -0,0 +1,394 @@
+{
+  "Description": "The CDK Toolkit Stack. It was created by `cdk bootstrap` and manages resources necessary for managing your Cloud Applications with AWS CDK.",
```
Comment (eladb) [inline packages/aws-cdk/lib/api/bootstrap/bootstrap-template.json]:
```suggestion
  "Description": "This stack includes resources needed to deploy AWS CDK apps into this environment",
```

File: packages/aws-cdk/lib/api/bootstrap/bootstrap-template.json [hunk 7]
```diff
@@ -0,0 +1,394 @@
...
+  "Parameters": {
+    "TrustedPrincipals": {
+      "Description": "List of AWS principals that the publish and action roles should trust to be assumed from",
```
Comment (eladb) [inline packages/aws-cdk/lib/api/bootstrap/bootstrap-template.json]: The description is unclear here. Maybe:

```suggestion
      "Description": "List of AWS principals that are trusted to publish assets and deploy stacks to this environment",
```

File: packages/aws-cdk/lib/api/bootstrap/bootstrap-template.json [hunk 8]
```diff
@@ -0,0 +1,394 @@
...
+      "Default": "",
+      "Type": "CommaDelimitedList"
+    },
+    "CloudFormationExecutionPolicies": {
+      "Description": "List of the ManagedPolicy ARN(s) to attach to the CloudFormation deployment role",
+      "Default": "",
+      "Type": "CommaDelimitedList"
+    },
+    "FileAssetsBucketName": {
+      "Description": "A user-provided custom name to use for the file assets bucket",
```
Comment (eladb) [inline packages/aws-cdk/lib/api/bootstrap/bootstrap-template.json]:
```suggestion
      "Description": "The name of the S3 bucket used for file assets",
```

File: packages/aws-cdk/lib/api/bootstrap/bootstrap-template.json [hunk 9]
```diff
@@ -0,0 +1,394 @@
...
+      "Default": "",
+      "Type": "String"
+    },
```
Comment (eladb) [inline packages/aws-cdk/lib/api/bootstrap/bootstrap-template.json]: Missing: custom name for ECR repository

File: packages/aws-cdk/lib/api/bootstrap/bootstrap-template.json [hunk 10]
```diff
@@ -0,0 +1,394 @@
...
+    "CustomKmsKeyId": {
```
Comment (eladb) [inline packages/aws-cdk/lib/api/bootstrap/bootstrap-template.json]:
```suggestion
    "FileAssetsBucketKmsKeyId": {
```

File: packages/aws-cdk/lib/api/bootstrap/bootstrap-template.json [hunk 11]
```diff
@@ -0,0 +1,394 @@
...
+      "Description": "A user-provided ID of a custom KMS key used for encrypting the file assets",
```
Comment (eladb) [inline packages/aws-cdk/lib/api/bootstrap/bootstrap-template.json]:
```suggestion
      "Description": "Custom KMS key ID to use for encrypting file assets (by default a KMS key will be automatically defined)",
```

File: packages/aws-cdk/lib/api/bootstrap/bootstrap-template.json [hunk 12]
```diff
@@ -0,0 +1,394 @@
...
+      "Default": "",
+      "Type": "String"
+    },
+    "CustomContainerAssetsRepositoryName": {
```
Comment (eladb) [inline packages/aws-cdk/lib/api/bootstrap/bootstrap-template.json]:
```suggestion
    "ContainerAssetsRepositoryName": {
```

File: packages/aws-cdk/lib/api/bootstrap/bootstrap-template.json [hunk 13]
```diff
@@ -0,0 +1,394 @@
+{
```
Comment (eladb) [inline packages/aws-cdk/lib/api/bootstrap/bootstrap-template.json]: Did we decide to add an optional resource name prefix (qualifier) or not?

File: packages/aws-cdk/lib/api/bootstrap/bootstrap-template.json (hunk 4 above)
Comment (eladb) [reply packages/aws-cdk/lib/api/bootstrap/bootstrap-template.json]: I think we should omit them.

Comment (skinny85) [reply packages/aws-cdk/lib/api/bootstrap/bootstrap-template.json]: The exports are required. See here: https://github.com/aws/aws-cdk/blob/master/design/cdk-bootstrap.md#cli-old-framework-old-bootstrap-new-init-template-old for details.

File: packages/aws-cdk/lib/api/bootstrap/bootstrap-template.json (hunk 13 above)
Comment (skinny85) [reply packages/aws-cdk/lib/api/bootstrap/bootstrap-template.json]: We decided not to do it in the bootstrap RFC.

File: packages/aws-cdk/lib/api/bootstrap/bootstrap-template.json (hunk 3 above)
Comment (skinny85) [reply packages/aws-cdk/lib/api/bootstrap/bootstrap-template.json]: This is just a semi-random string to reduce the chance of name conflicts.

File: packages/aws-cdk/lib/api/bootstrap/bootstrap-environment2.ts (hunk 5 above)
Comment (skinny85) [reply packages/aws-cdk/lib/api/bootstrap/bootstrap-environment2.ts]: I tried extracting a `deployStackFromTemplate()` function, but it was super awkward (the file are created differently (`fs.writeJson()` versus `fs.copy()`, the temporary directories have different hints, we need to reference the `output` property of the `CloudAssemblyBuilder` to know where to put the files, etc.), so I abandoned it for now. If you have any suggestions how to make that part better, I'm all ears.

I've added passing the CLI options through the parameters.

File: packages/aws-cdk/bin/cdk.ts (hunk 1 above)
Comment (skinny85) [reply packages/aws-cdk/bin/cdk.ts]: Changed to an env variable.

**Comment by skinny85** [comment]
@eladb @rix0rrr this is ready for a re-review.

File: packages/aws-cdk/bin/cdk.ts [hunk 14]
```diff
@@ -201,11 +204,17 @@ async function initCommandLine() {
         });
 
       case 'bootstrap':
-        return await cliBootstrap(args.ENVIRONMENTS, toolkitStackName, args.roleArn, {
+        // tslint:disable-next-line:no-console
```
Comment (eladb) [inline packages/aws-cdk/bin/cdk.ts]: Delete

File: packages/aws-cdk/lib/api/bootstrap/bootstrap-environment2.ts [hunk 15]
```diff
@@ -0,0 +1,45 @@
...
+import * as path from 'path';
+import { BootstrapEnvironmentProps, deployStack, DeployStackResult, ISDK } from '..';
+
+export async function bootstrapEnvironment2(environment: cxapi.Environment, sdk: ISDK,
+                                            toolkitStackName: string, roleArn: string | undefined,
+                                            props: BootstrapEnvironmentProps = {}): Promise<DeployStackResult> {
+  if (props.trustedAccounts?.length && !props.cloudFormationExecutionPolicies?.length) {
+    throw new Error('--cloudformation-execution-policies are required if --trust has been passed!');
+  }
+  // throw new Error('Break new bootstrap, just in case ;p');
```
Comment (eladb) [inline packages/aws-cdk/lib/api/bootstrap/bootstrap-environment2.ts]: ?

File: packages/aws-cdk/bin/cdk.ts (hunk 14 above)
Comment (skinny85) [reply packages/aws-cdk/bin/cdk.ts]: Removed.

File: packages/aws-cdk/lib/api/bootstrap/bootstrap-environment2.ts (hunk 15 above)
Comment (skinny85) [reply packages/aws-cdk/lib/api/bootstrap/bootstrap-environment2.ts]: Removed.

**Comment by skinny85** [comment]
Apologies for leaving the old comments in the PR @eladb , they are removed now.

File: packages/aws-cdk/lib/api/bootstrap/bootstrap-template.json (hunk 4 above)
Comment (rix0rrr) [reply packages/aws-cdk/lib/api/bootstrap/bootstrap-template.json]: But that mentions backwards compatibility with old CLIs, and the old CLI reads the Outputs, not necessarily the Exports. My question was around the exports.

File: packages/aws-cdk/lib/api/bootstrap/bootstrap-environment2.ts [hunk 16]
```diff
@@ -0,0 +1,44 @@
...
+import * as fs from 'fs-extra';
+import * as os from 'os';
+import * as path from 'path';
+import { BootstrapEnvironmentProps, deployStack, DeployStackResult, ISDK } from '..';
+
+export async function bootstrapEnvironment2(environment: cxapi.Environment, sdk: ISDK,
+                                            toolkitStackName: string, roleArn: string | undefined,
+                                            props: BootstrapEnvironmentProps = {}): Promise<DeployStackResult> {
+  if (props.trustedAccounts?.length && !props.cloudFormationExecutionPolicies?.length) {
+    throw new Error('--cloudformation-execution-policies are required if --trust has been passed!');
```
Comment (rix0rrr) [inline packages/aws-cdk/lib/api/bootstrap/bootstrap-environment2.ts]: I'm not super happy with this user experience. What if I'm trusting a new account? Not only do I have to list all previous accounts, but I also have to re-list the policy?

And the policy even needs to be referenced by ARN, which is super annoying.

I know we're going bare-bones functionality for now and worry about the experience later, but this is going to need some attention I feel...

File: packages/aws-cdk/lib/api/deploy-stack.ts [hunk 17]
```diff
@@ -54,6 +66,16 @@ export async function deployStack(options: DeployStackOptions): Promise<DeploySt
 
   const params = await prepareAssets(options.stack, options.toolkitInfo, options.reuseAssets);
 
+  // add passed CloudFormation parameters
+  for (const [paramName, paramValue] of Object.entries((options.parameters || {}))) {
+    if (paramValue) {
```
Comment (rix0rrr) [inline packages/aws-cdk/lib/api/deploy-stack.ts]: We can consider looking at the existing stack here, and if there are any Parameters on the current stack that aren't passed here, we can be passing `UsePreviousValue: true`  to make ergonomics around parameters a lot better.

Maybe not for this PR, but keep that in mind.  

https://docs.aws.amazon.com/AWSCloudFormation/latest/APIReference/API_Parameter.html

File: packages/aws-cdk/package.json [hunk 18]
```diff
@@ -39,6 +40,7 @@
   },
   "license": "Apache-2.0",
   "devDependencies": {
+    "@aws-cdk/core": "1.22.0",
```
Comment (rix0rrr) [inline packages/aws-cdk/package.json]: Danger Will Robinson! I don't think we want the CLI to depend on the framework, do we?

What is being used from this, and shouldn't it be in the cx-api instead?

File: packages/aws-cdk/test/api/bootstrap2.test.ts [hunk 19]
```diff
@@ -0,0 +1,56 @@
...
+  });
+
+  test('passes the KMS key ID as a CFN parameter', async () => {
+    await bootstrapEnvironment2(env, sdk, 'mockStack', undefined, {
+      kmsKeyId: 'my-kms-key-id',
+    });
+
+    expect(mockDeployStack).toHaveBeenCalled();
+
+    expect(mockDeployStackOptions).toMatchObject({
```
Comment (rix0rrr) [inline packages/aws-cdk/test/api/bootstrap2.test.ts]: You could have done

```
expect(mockDeployStack).toHaveBeenCalledWith({
      parameters: {
        FileAssetsBucketKmsKeyId: 'my-kms-key-id', 
     }
});
```

So you don't have to copy the options parameter out by hand.

File: packages/aws-cdk/test/integ/bootstrap/bootstrap.integ-test.ts [hunk 20]
```diff
@@ -0,0 +1,150 @@
...
+
+    describe('and then updates the bootstrap stack with the new resources', () => {
+      beforeAll(async () => {
+        // bootstrap the "new" way
+        const bootstrapResults = await bootstrapEnvironment2(env, sdk, bootstrapStackName, undefined, {
+          bucketName: newBootstrapBucketName,
+          trustedAccounts: ['790124522186', '593667001225'],
+          cloudFormationExecutionPolicies: [
+            'arn:aws:iam::aws:policy/AdministratorAccess',
+            'arn:aws:iam::aws:policy/AmazonS3FullAccess',
```
Comment (rix0rrr) [inline packages/aws-cdk/test/integ/bootstrap/bootstrap.integ-test.ts]: Is this not subsumed by AdministratorAccess?

File: packages/aws-cdk/test/integ/cli/test.sh [hunk 21]
```diff
@@ -5,6 +5,8 @@ source ${scriptdir}/common.bash
 
 header CLI Integration Tests
 
+npm run integ
```
Comment (rix0rrr) [inline packages/aws-cdk/test/integ/cli/test.sh]: I'd rather the execution of the js goes into its own shell script which runs just like the other shell scripts that represent tests, rather than being a different-class citizen like this.

File: packages/aws-cdk/lib/api/bootstrap/bootstrap-template.json (hunk 4 above)
Comment (skinny85) [reply packages/aws-cdk/lib/api/bootstrap/bootstrap-template.json]: Why do we have these Exports in the old bootstrapping?

File: packages/aws-cdk/lib/api/bootstrap/bootstrap-environment2.ts (hunk 16 above)
Comment (skinny85) [reply packages/aws-cdk/lib/api/bootstrap/bootstrap-environment2.ts]: I agree this is medium.

1. For the "trusting new accounts" problem: would your comment [here](https://github.com/aws/aws-cdk/pull/5856#discussion_r376365024) be applicable? We read the current value of the parameter, and add to it (with set semantics) the array of accounts that were passed in `--trust`, and that's the final value of the parameter? This way, `--trust` would always be additive.

2. For the "policy needs to be referenced by ARN" problem: originally, we wanted to prompt the customer if they didn't pass anything in the `--cloudformation-execution-policies` with a list like this:

```
This will create a new IAM Role, trusting any principal from the ACCOUNT_ID AWS account,
that will be used for deploying your CloudFormation stack(s) in that environment.
Please choose the Managed Policy or Policies to attach to the Role
that contain the permissions required to perform those deployments:

[1]  AdministratorAccess
[2]  AmazonDynamoDBFullAccess
[3]  AmazonEC2FullAccess
[4]  AmazonECS_FullAccess
[5]  AmazonS3FullAccess
[6]  AmazonSageMakerFullAccess
[7]  AmazonSESFullAccess
[8]  AmazonSNSFullAccess
[9]  AmazonSQSFullAccess
[10] AWSLambdaFullAccess
[11] SimpleWorkflowFullAccess
...

Please input the number of the chosen Policy from the above list,
or provide a fully qualified ARN of a different Managed Policy.
You can input multiple numbers and/or ARNs (up to 10),
separated by commas.

>
```

But it was scrapped for time purposes. Do you want to revisit this topic?

File: packages/aws-cdk/lib/api/deploy-stack.ts (hunk 17 above)
Comment (skinny85) [reply packages/aws-cdk/lib/api/deploy-stack.ts]: Great idea. I think we can use it to make `--trust` better - see https://github.com/aws/aws-cdk/pull/5856#discussion_r376673418 .

File: packages/aws-cdk/package.json (hunk 18 above)
Comment (skinny85) [reply packages/aws-cdk/package.json]: Relax Will. This is a devDependency needed only to compile the new test, written in TypeScript, which instantiates `App` from `@aws-cdk/core` for the test CDK app that is deployed using the newly created/updated bootstrap Stack. Here: https://github.com/aws/aws-cdk/pull/5856/files?file-filters%5B%5D=#diff-050637e62993e0e8bf23f8ad10fb636b

File: packages/aws-cdk/test/api/bootstrap2.test.ts (hunk 19 above)
Comment (skinny85) [reply packages/aws-cdk/test/api/bootstrap2.test.ts]: Nope:

```
Error: expect(jest.fn()).toHaveBeenCalledWith(...expected)

- Expected
+ Received

  Object {
+   "execute": undefined,
    "parameters": Object {
+     "CloudFormationExecutionPolicies": undefined,
      "FileAssetsBucketKmsKeyId": "my-kms-key-id",
+     "FileAssetsBucketName": undefined,
+     "TrustedAccounts": undefined,
+   },
+   "roleArn": undefined,
+   "sdk": MockSDK {
+     "credentialsCache": CredentialsCache {
+       "cache": Object {},
+       "defaultAwsAccount": DefaultAWSAccount {
+         "accountCache": AccountAccessKeyCache {
+           "cacheFile": "/Users/adamruka/.cdk/cache/accounts.json",
+         },
+         "defaultAccountFetched": false,
+         "defaultAccountId": undefined,
+         "defaultCredentialsProvider": Promise {},
+         "region": [Function anonymous],
+       },
+       "defaultCredentialProvider": Promise {},
+     },
+     "defaultAwsAccount": DefaultAWSAccount {
+       "accountCache": AccountAccessKeyCache {
+         "cacheFile": "/Users/adamruka/.cdk/cache/accounts.json",
+       },
+       "defaultAccountFetched": false,
+       "defaultAccountId": undefined,
+       "defaultCredentialsProvider": Promise {},
+       "region": [Function anonymous],
+     },
+     "profile": undefined,
+     "retryOptions": Object {
+       "maxRetries": 6,
+       "retryDelayOptions": Object {
+         "base": 300,
+       },
+     },
+     "sandbox": Sandbox {
+       "createStubInstance": [Function createStubInstance],
+       "fake": [Function fake],
+       "getFakes": [Function getFakes],
+       "getRestorers": [Function anonymous],
+       "inject": [Function inject],
+       "mock": [Function mock],
+       "replace": [Function replace],
+       "replaceGetter": [Function replaceGetter],
+       "replaceSetter": [Function replaceSetter],
+       "reset": [Function reset],
+       "resetBehavior": [Function resetBehavior],
+       "resetHistory": [Function resetHistory],
+       "restore": [Function restore],
+       "restoreContext": [Function restoreContext],
+       "serverPrototype": Object {
+         "addRequest": [Function addRequest],
+         "configure": [Function configure],
+         "create": [Function create],
+         "getHTTPMethod": [Function getHTTPMethod],
+         "getRequest": [Function getRequest],
+         "handleRequest": [Function handleRequest],
+         "log": [Function log],
+         "logError": [Function logError],
+         "logger": [Function logger],
+         "processRequest": [Function processRequest],
+         "reset": [Function reset],
+         "resetBehavior": [Function resetBehavior],
+         "resetHistory": [Function resetHistory],
+         "respond": [Function respond],
+         "respondAll": [Function respondAll],
+         "respondWith": [Function respondWith],
+         "restore": [Function restore],
        },
+       "spy": [Function spy],
+       "stub": [Function stub],
+       "useFakeServer": [Function useFakeServer],
+       "useFakeTimers": [Function useFakeTimers],
+       "useFakeXMLHttpRequest": [Function useFakeXMLHttpRequest],
+       "usingPromise": [Function usingPromise],
+       "verify": [Function verify],
+       "verifyAndRestore": [Function verifyAndRestore],
+     },
+   },
+   "stack": CloudFormationStackArtifact {
+     "_dependencyIDs": Array [],
+     "_deps": Array [],
+     "assembly": CloudAssembly {
+       "artifacts": Array [
+         [Circular],
+       ],
+       "directory": "/var/folders/gl/mhgz0kz11lqgc0dklyry83xnyfk1zw/T/cdk-bootstrap-new7TOYIv",
+       "manifest": Object {
+         "artifacts": Object {
+           "mockStack": Object {
+             "environment": "aws://123456789012/us-east-1",
+             "properties": Object {
+               "templateFile": "mockStack.template.json",
+             },
+             "type": "aws:cloudformation:stack",
+           },
+         },
+         "version": "1.21.0",
+       },
+       "runtime": Object {
+         "libraries": Object {},
+       },
+       "version": "1.21.0",
+     },
+     "assets": Array [],
+     "displayName": "mockStack",
+     "environment": Object {
+       "account": "123456789012",
+       "name": "aws://123456789012/us-east-1",
+       "region": "us-east-1",
+     },
+     "id": "mockStack",
+     "manifest": Object {
+       "environment": "aws://123456789012/us-east-1",
+       "properties": Object {
+         "templateFile": "mockStack.template.json",
+       },
+       "type": "aws:cloudformation:stack",
+     },
+     "messages": Array [],
+     "name": "mockStack",
+     "originalName": "mockStack",
+     "parameters": Object {},
+     "stackName": "mockStack",
+     "template": {
...
+     },
+     "templateFile": "mockStack.template.json",
+   },
+   "tags": undefined,
  },

Number of calls: 1

    at Object.test (/Users/adamruka/workplace/cdk/aws-cdk/packages/aws-cdk/test/api/bootstrap2.test.ts:32:33)
    at <anonymous>
```

File: packages/aws-cdk/test/integ/bootstrap/bootstrap.integ-test.ts (hunk 20 above)
Comment (skinny85) [reply packages/aws-cdk/test/integ/bootstrap/bootstrap.integ-test.ts]: Yeah, it is. Those roles are not used in the test in any way, I just wanted to test the case with more than one argument for the parameter. So I grabbed the first 2 managed policies that I could think of.

File: packages/aws-cdk/test/integ/cli/test.sh (hunk 21 above)
Comment (skinny85) [reply packages/aws-cdk/test/integ/cli/test.sh]: So, just to be 100% clear what needs to happen here:

1. You want to remove the above change.
2. You want to add a file in the `test/integ/cli` directory called something like `test-npm-integ.sh`, with the contents:

```bash
#!/bin/bash

set -euo pipefail
scriptdir=$(cd $(dirname $0) && pwd)
source ${scriptdir}/common.bash

npm run integ
```

?

File: packages/aws-cdk/lib/api/bootstrap/bootstrap-environment2.ts (hunk 16 above)
Comment (rix0rrr) [reply packages/aws-cdk/lib/api/bootstrap/bootstrap-environment2.ts]: > 1. For the "trusting new accounts" problem: would your comment here be applicable? We read the current value of the parameter, and add to it (with set semantics) the array of accounts that were passed in --trust, and that's the final value of the parameter? This way, --trust would always be additive.

Yes, I would definitely like that. We would only use `UsePreviousValue` if there was NO change to the parameter, but if we used `--trust` or `--untrust` (?) we would do set logic on the parameter.

2. I think I might want to revisit, or get people to the "right" solutions in an easier way. Maybe a UI, maybe we recognize some aliases for "standard" policies (`"admin" => "arn:aws:iam:::policy/AdministratorAccess"`). Doesn't have to be here and now (let's get this merged), but I want to put it into your head already.

File: packages/aws-cdk/package.json (hunk 18 above)
Comment (rix0rrr) [reply packages/aws-cdk/package.json]: Got it, thanks.

File: packages/aws-cdk/test/api/bootstrap2.test.ts (hunk 19 above)
Comment (rix0rrr) [reply packages/aws-cdk/test/api/bootstrap2.test.ts]: Let me rephrase that. You could have done:

```ts
expect(mockDeployStack).toHaveBeenCalledWith(expect.objectContaining({
      parameters: {
        FileAssetsBucketKmsKeyId: 'my-kms-key-id', 
     }
}));
```

(Maybe it needs an additional `expect.objectContaining()` around the `parameters` object)

File: packages/aws-cdk/test/integ/bootstrap/bootstrap.integ-test.ts (hunk 20 above)
Comment (rix0rrr) [reply packages/aws-cdk/test/integ/bootstrap/bootstrap.integ-test.ts]: ack

File: packages/aws-cdk/test/integ/cli/test.sh (hunk 21 above)
Comment (rix0rrr) [reply packages/aws-cdk/test/integ/cli/test.sh]: Indeed

File: packages/aws-cdk/lib/api/bootstrap/bootstrap-template.json (hunk 4 above)
Comment (skinny85) [reply packages/aws-cdk/lib/api/bootstrap/bootstrap-template.json]: Actually, it turns out we *don't* have those in the old bootstrapping 🤐.

That's weird - I could swear I copy-pasted this entire section from the existing bootstrap template, but I guess I didn't.

Will remove the exports.

File: packages/aws-cdk/lib/api/bootstrap/bootstrap-environment2.ts (hunk 16 above)
Comment (skinny85) [reply packages/aws-cdk/lib/api/bootstrap/bootstrap-environment2.ts]: > Yes, I would definitely like that. We would only use UsePreviousValue if there was NO change to the parameter, but if we used --trust or --untrust (?) we would do set logic on the parameter.

Blah. This is kind of hard ☹️.

The problem is that you can't use `UsePreviousValue` if you're creating the Stack. Which means it's not good enough for our case. We would have to call `DescribeStack` before the deployment, and based on what that returns:

* If the bootstrap Stack doesn't exist yet (= we're creating it), then pass whatever accounts were given in `--trust` as the value of the `TrustedAccounts` parameter.
* If the bootstrap Stack does exist, add the current value of the `TrustedAccounts` parameter (split on commas of course, as its Type is `CommaSeparatedList`) to the set of accounts passed through `--trust`, and use that as the value.

That also means we need some way to remove accounts from the trust relationship (like the `--untrust` CLI option you mentioned).

Do you want me to add this logic as part of this PR, or are you OK coming back to this later @rix0rrr ?

File: packages/aws-cdk/test/api/bootstrap2.test.ts (hunk 19 above)
Comment (skinny85) [reply packages/aws-cdk/test/api/bootstrap2.test.ts]: Nope, the first snippet does work, and it is better than what I had. Changed.

File: packages/aws-cdk/test/integ/cli/test.sh (hunk 21 above)
Comment (skinny85) [reply packages/aws-cdk/test/integ/cli/test.sh]: Done.

**Comment by skinny85** [comment]
@rix0rrr should be ready for another round of reviews (included your last comments).

File: packages/aws-cdk/lib/api/bootstrap/bootstrap-environment2.ts (hunk 16 above)
Comment (rix0rrr) [reply packages/aws-cdk/lib/api/bootstrap/bootstrap-environment2.ts]: Totally later!
